server/modules/playlist_driver.js

```diff
@@ -15,151 +15,181 @@ limitations under the License.
+    // as long as the current layout says to play modules.
+    let layout = this.playlist[this.layoutIndex];
+
+    // Stop any existing timer so we don't transition early.
+    // TODO(applmak): Consider making the timer management more foolproof by
+    // having the next* or play* methods stop the timer.
+    this.resetTimer_();
+
+    // Reset duration for this module.
+    this.newModuleTime = time.inFuture(layout.moduleDuration * 1000);
+    // Ensure that we won't change layouts until this module is done.
+    this.newLayoutTime = Math.max(this.newModuleTime, this.newLayoutTime);
+    // Now play this module.
+    this.playModule_(moduleName);
+  }
+  // Advances to the next layout in the playlist, fading out between them.
+  nextLayout() {
+    // Update layoutIndex.
+    this.layoutIndex = (this.layoutIndex + 1) % this.playlist.length;
+  
+    // Show this layout next:
+    let layout = this.playlist[this.layoutIndex];
   
-  moduleSM.setErrorListener(error => {
-    // Stop normal advancement.
-    clearTimeout(timer);
-    ret.nextModule();
-  });
+    // Reset moduleIndex
+    this.moduleIndex = -1;
+
+    // The time that we'll switch to a new layout.
+    this.newLayoutTime = time.inFuture(layout.duration * 1000);
+
+    if (monitor.isEnabled()) {
+      monitor.update({playlist: {
+        time: time.now(),
+        event: `change layout`,
+        deadline: this.newLayoutTime
+      }});
+    }
+
+    debug(`Next Layout: ${this.layoutIndex}`);
+
+    this.moduleSM.fadeToBlack(time.now() + 5000).then(() => {
+      // Shuffle the module list:
+      this.modules = Array.from(layout.modules);
+      random.shuffle(this.modules);
+
+      // Wait until all of the modules are loaded.
+      return Promise.all(layout.modules.map(m => m.whenLoadedPromise));
+    }).then(() => this.nextModule());
+  }
+  // Advances to the next module in the current layout. If there is only 1
+  // module in the current playlist, transitions to another copy of that
+  // module.
+  nextModule() {
+    this.moduleIndex = (this.moduleIndex + 1) % this.modules.length;
+
+    // The current layout.
+    let layout = this.playlist[this.layoutIndex];
   
-  return ret;
+    // The time that we'll switch to the next module.
+    this.newModuleTime = time.inFuture(layout.moduleDuration * 1000);
+
+    this.playModule_(this.modules[this.moduleIndex]);
+  }
+  // Private helper function that does the work of going to a module by name
+  // and scheduling the next module to play after a certain duration.
+  playModule_(module) {
+    // Play a module until the next transition time.
+    this.moduleSM.playModule(module, time.now());
+
+    if (monitor.isEnabled()) {
+      monitor.update({playlist: {
+        time: time.now(),
+        event: `change module ${module}`,
+        deadline: this.getNextDeadline(),
+      }});
+    }
+
+    // Now, in so many seconds, we'll need to switch to another module 
+    // or another layout. How much time do we have?
```

Comment:
I see "layout" mentioned all over in the code and comments. Weren't you working on removing that concept?